paper/src/main/resources/config.yml

```diff
@@ -428,38 +428,74 @@ hacks:
         - "We extend our appreciation to the devs, admins, and players at: &oreddit.com/r/Civcraft"
   InvControl:
     enabled: true
+
+  Limits:
+    enabled: false
+    chunk_limits:
+      cache_configuration:
+        # Capacity is elements, or complex objects, not bytes.
+        initial_capacity: 1000
+        maximum_capacity: 5000
+        expire_after_access_value: 15
+        expire_after_access_time_unit: MINUTES
+        # This will ensure that the entry is flushed after a certain time, to persist it.
+        expire_after_write_value: 2
+        expire_after_write_time_unit: HOURS
+
+      block_limits:
+# Limits are compounds. The following will allow at most 4 blocks of type piston and sticky piston in a chunk:
+#        - materials: [ PISTON, STICKY_PISTON ]
+#          max: 4
+# The next limit will limit the amount of pistons AND redstone blocks in a chunk:
+#        - materials: [ PISTON, REDSTONE_BLOCK ]
+#          max: 4
+# Both limits together allow at most 4 pistons, 4 sticky pistons and 4 red stone blocks.
+# These combinations are valid:
+# 1 piston, 3 sticky pistons, 3 redstone blocks.
+# 2 piston, 2 sticky pistons, 2 redstone blocks.
+# ...
+# 4 pistons, 0 sticky pistons, 0 redstone blocks.
+# 0 pistons, 4 sticky pistons, 4 redstone blocks
+# ...
+# Blocks in a list like this: [Stone, Glowstone] are summed and compared against the maximum.
+# For instance: [Stone, Glowstone] max: 1   will allow either a single stone, or a single glowstone.
+        - materials: [ PISTON, STICKY_PISTON ]
+          max: 4
+        - materials: [ PISTON, REDSTONE_BLOCK ]
+          max: 4
```

Comment:
After testing, you cannot place the max of 4 blocks. It prevents placement of the 4th block, rather than preventing placement of more than 4

Reply:
It appears this only occurs with certain blocks. For example, pistons and campfires have this issue, furnaces do not.

Reply:
https://github.com/psygate/SimpleAdminHacks/blob/9dd95e67c1f46fe1ae34f95054b7e7664e199fc1/paper/src/main/java/com/programmerdan/minecraft/simpleadminhacks/hacks/limits/Limits.java#L53

```
if (blockLimits.containsKey(mat)) {
	for (var limit : blockLimits.get(mat)) {
		var sum = sumOf(limit, chunkdata);
		if (sum >= limit.max()) {
			notifyPlayer(ev.getPlayer(), limit);
			ev.setCancelled(true);
			break;
		}
	}
}
```
I don't see how this is happening, and tested locally it works with set limits. You'll have to provide more information and more rigorous tests so I can triage the problem better.

Reply:
> It appears this only occurs with certain blocks. For example, pistons and campfires have this issue, furnaces do not.

I determined it is only specific entities. Pistons and campfires can have 1 less than the max placed. Furnaces can have the max placed.